staking/staking_hub.go

```diff
@@ -0,0 +1,394 @@
+		sh.EventQueue = append(sh.EventQueue[:foundIndx], sh.EventQueue[foundIndx+1:]...)
+		foundFlag = true
+	}
+
+	return foundFlag
+}
+
+// hasEvent checks if an identical event is present in the queue. Not thread safe
+func (sh *StakingHub) hasEvent(event PendingEvent) bool {
+	for _, el := range sh.EventQueue {
+		if el.Compare(event) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ContainsPendingEvent checks if an identical event is present in the queue
+func (sh *StakingHub) ContainsPendingEvent(event PendingEvent) bool {
+	sh.EventQueueMutex.Lock()
+	defer sh.EventQueueMutex.Unlock()
+
+	return sh.hasEvent(event)
+}
+
+// saveToDisk is a helper method for periodically saving the stake data to disk
+func (sh *StakingHub) saveToDisk() {
+	for {
+		select {
+		case <-sh.CloseCh:
+			return
+		default:
+		}
+
+		// Save the current staking map to disk, in JSON
+		mappings := sh.GetStakerMappings()
+
+		reader, err := sh.marshalJSON(mappings)
+		if err != nil {
+			continue
+		}
+
+		// Save the json to workingDirectory/StakingMap.json
+		file, err := os.Create(filepath.Join(sh.WorkingDirectory, defaultFileName))
+		if err != nil {
+			_ = file.Close()
+			sh.log("unable to create writeback file", logError)
+
+			continue
+		}
+
+		_, err = io.Copy(file, reader)
+		if err != nil {
+			sh.log("unable to write date into writeback file", logError)
+		}
+
+		err = file.Close()
+		if err != nil {
+			sh.log("unable to close writeback file", logError)
+		}
+
+		// Sleep for the writeback period
+		time.Sleep(sh.WritebackPeriod * time.Second)
+	}
+}
+
+// readFromDisk reads the staking map from a previously saved disk copy, if it exists
+func (sh *StakingHub) readFromDisk() error {
+	// Check if the file exists
+	mapFile, err := os.Open(filepath.Join(sh.WorkingDirectory, defaultFileName))
+	if err != nil {
+		return fmt.Errorf("no exiting map file is present")
+	}
+
+	byteValue, _ := ioutil.ReadAll(mapFile)
+
+	var stakingMap []StakerMapping
+
+	// Unmarshal the json
+	if err = json.Unmarshal(byteValue, &stakingMap); err != nil {
+		return fmt.Errorf("unable to unmarshal staking map")
+	}
+
+	// Close the file
+	if err = mapFile.Close(); err != nil {
+		return fmt.Errorf("unable to close staking file")
+	}
+
+	// Update the staking map using the read data
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	for _, val := range stakingMap {
+		sh.StakingMap[val.Address] = val.Stake
+	}
+
+	return nil
+}
+
+// marshalJSON generates the json object for staker mappings
+func (sh *StakingHub) marshalJSON(mappings []StakerMapping) (io.Reader, error) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	b, err := json.MarshalIndent(mappings, "", "\t")
+	if err != nil {
+		return nil, err
+	}
+	return bytes.NewReader(b), nil
+}
+
+// isStaker is a helper method to check whether or not an address has a staked balance
+func (sh *StakingHub) isStaker(address types.Address) bool {
+	if _, ok := sh.StakingMap[address]; ok {
+		return true
+	}
```

Comment:
Isn't it dangerous to only check if address is in the map like this?
As far as I see, increasing stake and decreasing it by the same balance will still leave the address in the staking map resulting in this method returning true

Reply:
The only places where we modify the staking map directly are through methods `IncreaseStake`, which adds the address to the map, `DecreaseStake`(used only in tests) and `ResetStake`, which deletes that entry in the staking map:
https://github.com/0xPolygon/polygon-sdk/blob/c2e920b30aa55937496816e098d90b57adafc5ca/staking/staking_hub.go#L332-L334

GIven this, we can be sure that if the stake was reset, the address will not be in the map

Reply:
What happens if some address increases its stake by 10, then decreases it by 10?
It will be in the`StakingMap`, and `isStaker` will return `true` for that address, but `false` for an address that never staked anything.

I find that a bit misleading for anyone using this method, especially given it's description:
```
// isStaker is a helper method to check whether or not an address has a staked balance
```

Reply:
@zivkovicmilos Is this still unresolved?

Reply:
I've changed up the description so it is more clear as to what the function actually does 👍 :
https://github.com/0xPolygon/polygon-sdk/blob/50f5bac2b7acc2e78ca446f429d6c47b0330e4a3/staking/staking_hub.go#L266-L275

Reply:
I think the description isn't enough given that this function changes behavior depending on what txs were executed in the block prior to the call. Can we make it a bit more unambiguous, like this:
```suggestion
	if _, ok := sh.StakingMap[address]; ok {
		return sh.StakingMap[address].Cmp(big.NewInt(0)) > 0
	}
```

Reply:
Got it.
Fixed 👍 

37e054c